Change finder presenter logic for nested facets

lauraghiorghisor-tw opened the pull request:
Changing nested facets implementation. We now send two distinct facets in the publishing-api load, corresponding to the main and sub facet, respectively, in order to minimise how much we need to change Finder Frontend down the line.

[Trello](https://trello.com/c/skH2B2fH/3373-nested-facets-finder-frontend)

lauraghiorghisor-tw committed:
Update prepositions for trademark decisions

minhngocd commented on app/presenters/finder_content_item_presenter.rb:
```diff
@@ -60,12 +60,57 @@ def details
       signup_link: file.fetch("signup_link", nil),
       summary: file.fetch("summary", nil),
       label_text: file.fetch("label_text", nil),
-      facets: FinderContentItemPresenter.facets_without_specialist_publisher_properties(file.fetch("facets", nil)),
```
should we pull this out as a `FinderFacetPresenter` ?

minhngocd commented on app/presenters/finder_content_item_presenter.rb:
```diff
@@ -60,12 +60,57 @@ def details
+
+  def main_facet_hash_without_sub_facets(facet_hash)
+    hash_dup = facet_hash.dup
+    hash_dup["allowed_values"] = hash_dup["allowed_values"].map { |v| v.except("sub_facets") }
```
can we not use hash.except(key)?

minhngocd commented on app/presenters/finder_content_item_presenter.rb:
```diff
@@ -60,12 +60,57 @@ def details
+  def sub_facet_allowed_values(facet_hash)
+    sub_facet_values_with_main_facet_reference = []
+
+    facet_hash["allowed_values"].each do |allowed_value|
```
instance of inject to replace the `[]` instantiation and return

ChrisBAshton commented on app/presenters/finder_content_item_presenter.rb:
```diff
@@ -27,14 +27,6 @@ def content_id
     file.fetch("content_id")
   end
 
-  def self.facets_without_specialist_publisher_properties(facets)
```
It would be good to do the moving of this into FinderFacetPresenter in a separate commit to then adding the sub_facet_hash logic.

minhngocd replied:
Updated now. The refactoring is in a separate commit

minhngocd committed:
Refactoring: Split FinderFacetPresenter out of FinderContentItemPresenter

There is going to be additional logic we want to add into finder facets, which makes sense for us to pull into a separate object.
The new object allows us to remove the static method currently there, and simplifies testing for facets.

minhngocd committed:
Map single nested facet from schema to two individual facets for publishing-api

In order to minimise the customisation in Finder Frontend, we will send the data in the shape it is typically expected by Finder Frontend, i.e. the facets are stand-alone. We still need some way of referencing the parent facet value in the children, which we're doing by setting `main_facet_value`, alongside every subfacet label-value pair in the subfacet. We also send the `main_facet_label` so we can recompose the subfacet label in the format `Main facet label - Sub facet label` when JS is disabled and all the facet options show in the select; since we have duplicates such as "Not applicable" it would be unclear which main facet option the subfacet option refers to.

Other required keys are:
- the `nested_facet` (T/F) flag which indicates to the Finder Frontend rendering app that the facet should be rendered as a NestedFacet (new Facet type added by us).
- `sub_facet_key` and `sub_facet_name` which are only set on the main facet, used for loading the js module.
- the `key` and `name` of the subfacet come from the data stored under the main facet, in the schema.
- all the other key values are shared between the main and sub facets, though a case could be made in the future for also allowing the customization of the subfacet's `short_name`, `preposition`, `display_as_result_metadata` or `show_option_select_filter` values.

Co-authored-by: Laura Ghiorghisor <laura.ghiorghisor@digital.cabinet-office.gov.uk>

lauraghiorghisor-tw committed:
Lowercase the document noun for trademark decisions
